cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+	kA := c.suite.Point().Mul(minusPublic, k)
+	sB := c.suite.Point().Mul(nil, c.aggregateResponse)
+	left := c.suite.Point().Add(kA, sB)
+
+	if !left.Equal(c.aggregateCommitment) {
+		return errors.New("recreated commitment is not equal to one given")
+	}
+
+	return nil
+}
+
+// VerifySignature is the method to call to verify a signature issued by a Cosi
+// struct. Publics is the WHOLE list of publics keys, the mask at the end of the
+// signature will take care of removing the indivual public keys that did not
+// participate
+func VerifySignature(suite abstract.Suite, publics []abstract.Point, message, sig []byte) error {
+	aggCommitBuff := sig[:32]
+	aggCommit := suite.Point()
+	if err := aggCommit.UnmarshalBinary(aggCommitBuff); err != nil {
+		panic(err)
+	}
+	sigBuff := sig[32:64]
+	sigInt := sliceToSecret(suite, sigBuff)
+	maskBuff := sig[64:]
+	mask := NewCosiMask(suite, publics)
+	mask.Set(maskBuff)
+	aggPublic := mask.Aggregate()
+	aggPublicMarshal, err := aggPublic.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	hash := sha512.New()
+	hash.Write(aggCommitBuff)
+	hash.Write(aggPublicMarshal)
+	hash.Write(message)
+	kBuff := hash.Sum(nil)
+	k := sliceToSecret(suite, kBuff)
+
+	// k * -aggPublic + s * B = k*-A + s*B
+	// from s = k * a + r => s * B = k * a * B + r * B <=> s*B = k*A + r*B
+	// <=> s*B + k*-A = r*B
+	minusPublic := suite.Point().Neg(aggPublic)
+	kA := suite.Point().Mul(minusPublic, k)
+	sB := suite.Point().Mul(nil, sigInt)
+	left := suite.Point().Add(kA, sB)
+
+	if !left.Equal(aggCommit) {
+		return errors.New("Signature invalid")
+	}
+
+	return nil
+}
+
+// genCommit generates a random secret vi and computes it's individual commit
+// Vi = G^vi
+func (c *Cosi) genCommit(r io.Reader) {
+	var reader = r
+	if r == nil {
+		reader = cryptoRand.Reader
+	}
+	var randomFull [64]byte
+	if _, err := io.ReadFull(reader, randomFull[:]); err != nil {
+		panic(err)
+	}
+	c.random = sliceToSecret(c.suite, randomFull[:])
+	c.commitment = c.suite.Point().Mul(nil, c.random)
+	c.aggregateCommitment = c.commitment
+}
+
+// genResponse creates the response
+func (c *Cosi) genResponse() error {
+	if c.private == nil {
+		return errors.New("No private key given in this cosi")
+	}
+	if c.random == nil {
+		return errors.New("No random secret computed in this cosi")
+	}
+	if c.challenge == nil {
+		return errors.New("No challenge computed in this cosi")
+	}
+
+	// hash the private key
+	hash := sha512.New()
+	privKeyBuff := secretToSlice(c.private)
+	//privKeyBuff := c.private.(*nist.Int).LittleEndian(32, 32)
+
+	hash.Write(privKeyBuff)
+	h := hash.Sum(nil)
+
+	// prune it up
+	expandedPrivKey := h[0:32]
+	expandedPrivKey[0] &= 248
+	expandedPrivKey[31] &= 127
+	expandedPrivKey[31] |= 64
+	expandedPrivKeyInt := sliceToSecret(c.suite, expandedPrivKey)
+
+	// resp = random - challenge * privatekey
+	// i.e. ri = vi + c * xi
+	resp := c.suite.Secret().Mul(expandedPrivKeyInt, c.challenge)
+	c.response = resp.Add(c.random, resp)
```

Comment:
Ok since this is the second argument regarding whether we should send the individual response or not in the message, I'll remove the *individual response* in the packets and then I'll be able to put the response to nil here also.